libs/pika/async_mpi/tests/unit/CMakeLists.txt

```diff
@@ -4,25 +4,42 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-set(tests algorithm_transform_mpi mpi_ring_async_sender_receiver)
+set(tests algorithm_transform_mpi mpi_ring_async_sender_receiver pool_creation)
 
 # cmake-format: off
 set(mpi_ring_async_sender_receiver_PARAMETERS
-    ARGS "--in-flight-limit=32" "--rounds=5" "--iterations=25" "--message-bytes=64"
-         "--pika:ignore-process-mask"
-    THREADS 4 RANKS 2 MPIWRAPPER
+    ARGS
+      "--in-flight-limit=2"
+      "--rounds=5"
+      "--iterations=25"
+      "--message-bytes=64"
+      "--pika:ignore-process-mask"
+    THREADS 4
+    RANKS 2
+    MPIWRAPPER
 )
 
 set(mpi_async_storage_PARAMETERS
-    ARGS "--in-flight-limit=256" "--localMB=256" "--transferKB=1024" "--seconds=1" 
-         "--pika:ignore-process-mask"
-    THREADS 4 RANKS 2 MPIWRAPPER
+    ARGS
+      "--in-flight-limit=256"
+      "--localMB=256"
+      "--transferKB=1024"
+      "--seconds=1"
+      "--pika:ignore-process-mask"
+    THREADS 4
+    RANKS 2
+    MPIWRAPPER
 )
-# cmake-format: on
 
 set(algorithm_transform_mpi_PARAMETERS THREADS 2 RANKS 2 MPIWRAPPER)
 set(algorithm_transform_mpi_DEPENDENCIES pika_execution_test_utilities)
 
+set(pool_creation_PARAMETERS THREADS 2 RANKS 2 MPIWRAPPER)
+set(pool_creation_NO_POOL ON)
+# cmake-format: on
+
+# note that because we add (eg) "_0_5" to the test name, we explicitly link pika_testing because
+# pika_add_unit_test expects "${name}_test"
```

Comment:
I don't understand this comment. On main we currently name the executable (e.g.) `mpi_ring_async_sender_receiver_test` and the tests are called `mpi_ring_async_sender_receiver_mode_X_test` and it links just fine. `pika_add_test` will link any executable ending with `_test` to `pika_testing` (https://github.com/pika-org/pika/blob/34f044e33b242a87ca1b2a06a702e0763af11087/cmake/pika_add_test.cmake#L108-L110), and as far as I can tell you haven't changed the executable name.

Reply:
we pass in a name of (eg) `mpi_ring_async_sender_receiver_mode_0_0` and it has flags `EXECUTABLE ${test}_test` which is correct (the test executable is `mpi_ring_async_sender_receiver_test`, but the check you linked to tests
```
if(TARGET ${${name}_EXECUTABLE}_test 
```
which will expand to `mpi_ring_async_sender_receiver_test_test` which is not correct and I couldn't be arsed fixing it, so just manually linked the testing lib.
Now that I write it down, I could fix it by just passing the EXECUTABLE as `mpi_ring_async_sender_receiver` and then the `test` suffix will match. I will try that ...

Reply:
well, that does work, but it strikes me as being wrong. the test is added with 
```
  pika_add_executable(
    ${test}_test 
```
which has `_test` appended as a suffix, so the executable is called xxx_test, so passing ```EXECUTABLE xxx_test```  syntactically seems right and the add_unit_test adding `_test` again is unnecessary.  I hope this makes sense. it is a bit confusing, anyway, I can either link `pika_testing` as I'm doing or use the (now) modified version of
```
      pika_add_unit_test(
        "modules.async_mpi"
        ${full_name_}
        ${${test}_PARAMETERS}
        EXECUTABLE
        ${test}
        ARGS
        $<$<BOOL:${enable_pool}>:--pika:mpi-enable-pool>
        "--pika:mpi-completion-mode=${polling_mode}"
      )
```
which drops the test suffix and works.

Reply:
(what I forgot to say last night is that, really add_unit_test, is just adding a wrapper around an executable to invoke it with the right params, and the linking to pika_testing should happen in the add_executable itself - which is where I added it in the PR originally). But the comment was completely wrong :(

Reply:
That's a good catch! However, given that it works here without the changes you've made could you remove these changes from here? I've opened https://github.com/pika-org/pika/pull/1271 as a starting point to fix it and/or change the behaviour.

I think I agree in one sense that adding a test should not link. On the other hand this is done to enforce that tests and only tests link to `pika_testing`. If it's a test it should always be linking to it (and fail if no tests were actually done with `PIKA_TEST` et al., since we fail a test if no tests were run). If it's not a test it shouldn't link to `pika_testing`. And if it's using `PIKA_TEST` but pretending it's not a test, it is a test.

I'd like to avoid that we have to explicitly link to `pika_testing` everywhere. A half-way explicit option could be to add a `TESTING` flag to `pika_add_executable`. Would you mind commenting on https://github.com/pika-org/pika/pull/1271 if you have a preference or have other ideas on how we could do it differently?

Reply:
An executable either needs to link to a library or not - whether you actually use it as a unit test or not should not affect the link rules (unless there is some kind of plugin mechanism at play). I did not look at what `pika_testing` actually provides but I will add this to my todo list

Reply:
I half-agree.

Forcing the linking of `pika_testing` is to ensure that a unit test actually does some testing. A unit test that doesn't use `PIKA_TEST` is either not a unit test or someone forgot to use `PIKA_TEST`. Conversely, a test that pretends not to be a unit test but uses `PIKA_TEST` is a unit test, whether it wants it or not. In other words, these all imply each other "link to `pika_testing`" <=> "uses `PIKA_TEST`" <=> "is a unit test".

No plugin mechanism at play.

We do this at the exit of the test fixture: https://github.com/pika-org/pika/blob/4bb041ca52bdf61ab541f49830604e7291f04c41/testing/testing/src/testing.cpp#L37, in case you want to dig into it further.

I'll mark this as resolved though since this part is now unchanged on this PR.